### PR TITLE
Use DarksidewalletD tests on `zingocli`

### PR DESCRIPTION
**Pull request by pacu**
This is PoC of how zingocli could leverage darksidewalletd to make advanced reorg tests as ZcashLightClientKit does.

How to run darksidewalletd

clone lightwalletd from Git from branch `master`
`make` the project. 

run this script to launch a local lwd in darkside mode.

````Bash
#!/bin/zsh

./lightwalletd --log-file /dev/stdout --darkside-very-insecure  --darkside-timeout 1000 --gen-cert-very-insecure --data-dir . --no-tls-very-insecure
````

on `zingocli` run `sh tests/scripts/start_darksidewalletd.sh` to set up the state of darksidewalletd

then run this test

image

expected: the wallet should sync up to block 663200
result: it does not (see comment)

### CODE REVIEW

**Comment by pacu** [comment]
Currently there's an issue with checkpoints that have no orchard tree value (pre NU5)

````
running 1 test
thread 'test_simple_sync' panicked at 'byte index 132 is out of bounds of `000000`', zingolib/src/blaze/block_witness_data.rs:337:20
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:65:14
   2: core::str::slice_error_fail_rt
   3: core::str::slice_error_fail
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/str/mod.rs:86:9
   4: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::RangeTo<usize>>::index
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/str/traits.rs:294:21
   5: core::str::traits::<impl core::ops::index::Index<I> for str>::index
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/str/traits.rs:65:9
   6: <alloc::string::String as core::ops::index::Index<core::ops::range::RangeTo<usize>>>::index
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/alloc/src/string.rs:2371:10
   7: zingolib::blaze::block_witness_data::BlockAndWitnessData::verify_tree_pair::{{closure}}
             at /Users/pacu/Repos/ECC/zingolib/zingolib/src/blaze/block_witness_data.rs:337:20
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/future/mod.rs:91:19
   9: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/core.rs:223:17
  10: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/loom/std/unsafe_cell.rs:14:9
  11: tokio::runtime::task::core::Core<T,S>::poll
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/core.rs:212:13
  12: tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/harness.rs:476:19
  13: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panic/unwind_safe.rs:271:9
  14: std::panicking::try::do_call
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:483:40
  15: ___rust_try
  16: std::panicking::try
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:447:19
  17: std::panic::catch_unwind
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panic.rs:137:14
  18: tokio::runtime::task::harness::poll_future
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/harness.rs:464:18
  19: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/harness.rs:198:27
  20: tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/harness.rs:152:15
  21: tokio::runtime::task::raw::poll
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/raw.rs:255:5
  22: tokio::runtime::task::raw::RawTask::poll
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/raw.rs:200:18
  23: tokio::runtime::task::LocalNotified<S>::run
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/task/mod.rs:394:9
  24: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:584:25
  25: tokio::runtime::coop::with_budget
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/coop.rs:107:5
  26: tokio::runtime::coop::budget
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/coop.rs:73:5
  27: tokio::runtime::scheduler::current_thread::Context::run_task::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:285:29
  28: tokio::runtime::scheduler::current_thread::Context::enter
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:350:19
  29: tokio::runtime::scheduler::current_thread::Context::run_task
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:285:9
  30: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:583:34
  31: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:615:57
  32: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/macros/scoped_tls.rs:61:9
  33: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:615:27
  34: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:530:19
  35: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:154:24
  36: tokio::runtime::runtime::Runtime::block_on
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/runtime.rs:302:47
  37: integration_tests::test_simple_sync
             at ./tests/integration_tests.rs:1134:5
  38: integration_tests::test_simple_sync::{{closure}}
             at ./tests/integration_tests.rs:1112:7
  39: core::ops::function::FnOnce::call_once
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:251:5
  40: core::ops::function::FnOnce::call_once
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'test_simple_sync' panicked at 'called `Result::unwrap()` on an `Err` value: "Tree Verification Failed"', zingocli/tests/integration_tests.rs:1125:45
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:65:14
   2: core::result::unwrap_failed
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/result.rs:1791:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/result.rs:1113:23
   4: integration_tests::test_simple_sync::{{closure}}
             at ./tests/integration_tests.rs:1125:18
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/future/mod.rs:91:19
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/future/future.rs:124:9
   7: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/future/future.rs:124:9
   8: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:541:57
   9: tokio::runtime::coop::with_budget
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/coop.rs:107:5
  10: tokio::runtime::coop::budget
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/coop.rs:73:5
  11: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:541:25
  12: tokio::runtime::scheduler::current_thread::Context::enter
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:350:19
  13: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:540:36
  14: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:615:57
  15: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/macros/scoped_tls.rs:61:9
  16: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:615:27
  17: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:530:19
  18: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/current_thread.rs:154:24
  19: tokio::runtime::runtime::Runtime::block_on
             at /Users/pacu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/runtime.rs:302:47
  20: integration_tests::test_simple_sync
             at ./tests/integration_tests.rs:1134:5
  21: integration_tests::test_simple_sync::{{closure}}
             at ./tests/integration_tests.rs:1112:7
  22: core::ops::function::FnOnce::call_once
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:251:5
  23: core::ops::function::FnOnce::call_once
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test test_simple_sync ... FAILED

failures:

failures:
    test_simple_sync

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 19 filtered out; finished in 0.55s

error: test failed, to rerun pass `-p zingo-cli --test integration_tests`

 *  The terminal process "cargo 'test', '--package', 'zingo-cli', '--test', 'integration_tests', '--', 'test_simple_sync', '--exact', '--nocapture'" terminated with exit code: 101. 
 *  Terminal will be reused by tasks, press any key to close it. 
 ````

**Comment by pacu** [comment]
> This is exciting, thanks @pacu !

 thanks!
 
 I think the bug is worth looking. I don't think its related to the PR itself but to inherent to the code it exercises

**Comment by pacu** [comment]
@zancas @AloeareV I added a test that actually reproduces the bug I'm describing in the comment above. 

 block\_witness\_date.rs -> verify\_tree\_pair() function, there's this if statement that fails when it receives tree states pre-orchard.

```
// Return if verified
            if (hex::encode(sapling_buf) == unverified_tree.sapling_tree)
                && (determined_orchard_tree == unverified_tree.orchard_tree)
            {
                true
            } else {
                // Parentless trees are encoded differently by zcashd for Orchard than for Sapling
                // this hack allows the case of a parentless orchard_tree
                if determined_orchard_tree[..132] == unverified_tree.orchard_tree[..132]
                    && &determined_orchard_tree[132..] == "00"
                    && unverified_tree.orchard_tree[134..]
                        .chars()
                        .all(|character| character == '0')
                {
                    true
                } else {
                    dbg!(determined_orchard_tree, unverified_tree.orchard_tree);
                    false
                }
            }
```

determined\_orchard\_tree is "000000", so slicing it like this `determined_orchard_tree[..132]` makes the thing blow up in the air and panic

**Comment by pacu** [comment]
the way to reproduce the issue is just by running the test

`cargo test --package zingolib --lib -- blaze::block_witness_data::test::test_verified_parentless_orchard_tree --exact --nocapture `

File: zingocli/src/lib.rs [hunk 1]
```diff
@@ -13,6 +13,8 @@ use zingolib::{commands, create_zingoconf_from_datadir, lightclient::LightClient
 pub mod regtest;
 pub mod version;
 
+
+
```
Comment (zancas) [inline zingocli/src/lib.rs:17]: Shall we remove this change?

File: zingocli/tests/darksidewalletd_tests.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (zancas) [inline zingocli/tests/darksidewalletd_tests.rs:1]: I like the new test file.  These are a fundamentally new type of test..   why not give them their own file?

File: zingolib/src/lightclient/checkpoints.rs [hunk 3]
```diff
@@ -27,7 +27,7 @@ fn get_test_checkpoint(height: u64) -> Option<(u64, &'static str, &'static str)>
 
 /// Note that all checkpoints are pre-orchard activation, and their fields are:
 /// height, hash, sapling_tree. Will need to be restructured when post-orchard points are added
-pub const fn get_all_main_checkpoints() -> [(u64, &'static str, &'static str); 21] {
+pub const fn get_all_main_checkpoints() -> [(u64, &'static str, &'static str); 22] {
```
Comment (zancas) [inline zingolib/src/lightclient/checkpoints.rs:30]: It seems like we should update the doc-comment associated with this `fn` also.

File: zingolib/src/blaze/block_witness_data.rs [hunk 4]
```diff
@@ -827,6 +817,20 @@ impl BlockAndWitnessData {
     }
 }
 
+fn is_orchard_tree_verified(determined_orchard_tree: String, unverified_tree: TreeState) -> bool {
```
Comment (zancas) [inline zingolib/src/blaze/block_witness_data.rs:820]: I really like this.   I think this helperization/test/bug reproduction belongs in its own PR.  Should we split it out?

File: zingocli/tests/darksidewalletd_tests.rs (hunk 2 above)
Comment (pacu) [reply zingocli/tests/darksidewalletd_tests.rs:1]: I'm not done with that. Tests wouldn't compile for some reason that is probably obvious for a real Rust dev but not to me 😅

File: zingolib/src/blaze/block_witness_data.rs (hunk 4 above)
Comment (pacu) [reply zingolib/src/blaze/block_witness_data.rs:820]: good Idea. I'll split it out and send a new PR with that change.

Comment (pacu) [reply zingolib/src/blaze/block_witness_data.rs:820]: @zancas  I created a PR for this! #298

**Comment by pacu** [comment]
I just pulled the rebased branch from @zancas repo and forced pushed here. 

Thanks @zancas

**Comment by pacu** [comment]
@zancas I pushed your rebased changes (thank you!) and fixed some small things. I was able to run the darksidewalletd test successfully

**Comment by zancas** [comment]
Hey @AloeareV is there still unique contribs to capture here?   Or is this work already merged elsewhere?

**Comment by zancas** [comment]
Hi!   This (excellent) work has been merged in other PRs.

**Pull request closed without merging**